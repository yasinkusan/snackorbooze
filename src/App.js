import React, { useState, useEffect } from 'react';
import './App.css';
import Home from './Home';
import SnackOrBoozeApi from './Api';
import NavBar from './NavBar';
import { Route, Switch } from 'react-router-dom';
import Menu from './FoodMenu';
import Item from './FoodItem';
import NewItemForm from './NewItemForm';

function App() {
	const [ isLoading, setIsLoading ] = useState(true);
	const [ snacks, setSnacks ] = useState([]);
	const [ drinks, setDrinks ] = useState([]);

	// get drinks and snacks on page load and set state accordingly.
	useEffect(() => {
		async function getItems() {
			let snackList = await SnackOrBoozeApi.getSnacks();
			let drinkList = await SnackOrBoozeApi.getDrinks();
			setSnacks(snackList);
			setDrinks(drinkList);
			setIsLoading(false);
		}
		getItems();
	}, []);

	// function to add new item to state with proper formatting to match the db, and add to the db based on type
	const addNewItem = async (newItem) => {
		let itemFormatted = {
			...newItem,
			id      : newItem.name.toLowerCase().replace(' ', '-'),
			userAdd : true
		};
		// logic to decide whether to add to snacks or drinks state/db
		if (newItem.type === 'snack') {
			await SnackOrBoozeApi.addSnack(itemFormatted);
			setSnacks((snacks) => [ ...snacks, itemFormatted ]);
		} else if (newItem.type === 'drink') {
			await SnackOrBoozeApi.addDrink(itemFormatted);
			setDrinks((drinks) => [ ...drinks, itemFormatted ]);
		}
	};

	// function to delete an item off the menu if it was user-added
	const deleteItem = async (item) => {
		if (item.type === 'snack') {
			setSnacks((snacks) => snacks.filter((snack) => snack.id !== item.id));
			await SnackOrBoozeApi.deleteSnack(item.id);
		} else if (item.type === 'drink') {
			setDrinks((drinks) => drinks.filter((drink) => drink.id !== item.id));
			await SnackOrBoozeApi.deleteDrink(item.id);
		}
	};

	if (isLoading) {
		return <p style={{ color: 'white' }}>Loading &hellip;</p>;
	}

	return (
		<div className="App">
			<NavBar />
			<main>
				<Switch>
					<Route exact path="/">
						<Home snacks={snacks} drinks={drinks} />
					</Route>
					<Route exact path="/snacks">
						<Menu items={snacks} title="Snacks" remove={deleteItem} />
					</Route>
					<Route path="/snacks/:id">
						<Item items={snacks} cantFind="/snacks" backTo="snacks" />
					</Route>
					<Route exact path="/drinks">
						<Menu items={drinks} title="Drinks" remove={deleteItem} />
					</Route>
					<Route exact path="/DRINKS/:id">
						<Item items={drinks} cantFind="/drinks" backTo="drinks" />
					</Route>
					<Route exact path="/new">
						<NewItemForm add={addNewItem} toggleLoad={setIsLoading} />
					</Route>
					<Route>
						<h2 className="App-notFound">Hmmm. I can't seem to find what you want.</h2>
					</Route>
				</Switch>
			</main>
		</div>
	);
}

export default App;
