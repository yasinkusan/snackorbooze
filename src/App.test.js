import React from 'react';
import App from './App';
import { MemoryRouter } from 'react-router-dom';

import { render, fireEvent } from '@testing-library/react';
import { act } from 'react-dom/test-utils';

function sleep(time) {
	return new Promise((resolve) => setTimeout(resolve, time));
}

it('renders without crashing', () => {
	render(
		<MemoryRouter initialEntries={[ '/' ]}>
			<App />
		</MemoryRouter>
	);
});

it('matches loading snapshot', () => {
	const { asFragment } = render(
		<MemoryRouter initialEntries={[ '/' ]}>
			<App />
		</MemoryRouter>
	);
	expect(asFragment()).toMatchSnapshot();
});
it('matches rendered snapshot', async () => {
	const { asFragment } = render(
		<MemoryRouter initialEntries={[ '/' ]}>
			<App />
		</MemoryRouter>
	);
	await act(() => sleep(1100));
	expect(asFragment()).toMatchSnapshot();
});

it('can render the /home route', async () => {
	const { getByText } = render(
		<MemoryRouter initialEntries={[ '/' ]}>
			<App />
		</MemoryRouter>
	);

	await act(() => sleep(1100));
	const text = getByText("Welcome to Silicon Valley's premier dive cafe!");
	expect(text).toBeInTheDocument();
});

it('can use navbar links', async () => {
	const { getByText } = render(
		<MemoryRouter initialEntries={[ '/' ]}>
			<App />
		</MemoryRouter>
	);

	await act(() => sleep(1100));
	const navLink = getByText('Snacks');
	expect(navLink).toBeInTheDocument();
	fireEvent.click(navLink);
	expect(navLink).toBeInTheDocument();
	expect(getByText('Explore all of our delish snacks below!')).toBeInTheDocument();
});

it('can use in-line links', async () => {
	const { getByText } = render(
		<MemoryRouter initialEntries={[ '/' ]}>
			<App />
		</MemoryRouter>
	);

	await act(() => sleep(1100));
	const link = getByText('3 snacks');
	expect(link).toBeInTheDocument();
	fireEvent.click(link);
	expect(getByText('Explore all of our delish snacks below!')).toBeInTheDocument();
	expect(link).not.toBeInTheDocument();
});
it('can render a snack element', async () => {
	const { getByText, getByTestId } = render(
		<MemoryRouter initialEntries={[ '/' ]}>
			<App />
		</MemoryRouter>
	);

	await act(() => sleep(1100));
	const link = getByText('3 snacks');
	console.log(link);
	expect(link).toBeInTheDocument();
	fireEvent.click(link);
	expect(getByText('Explore all of our delish snacks below!')).toBeInTheDocument();
	expect(link).not.toBeInTheDocument();
	const newLink = getByTestId('link-hummus');
	expect(newLink).toBeInTheDocument();
	fireEvent.click(newLink);
	expect(newLink).not.toBeInTheDocument();
	expect(getByText('Sure to impress your vegan friends!')).toBeInTheDocument();
});
