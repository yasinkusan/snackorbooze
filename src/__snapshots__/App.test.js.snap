// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches loading snapshot 1`] = `
<DocumentFragment>
  <p
    style="color: white;"
  >
    Loading â€¦
  </p>
</DocumentFragment>
`;

exports[`matches rendered snapshot 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <div>
      <nav
        class="navbar navbar-expand-md"
      >
        <a
          aria-current="page"
          class="navbar-brand active"
          href="/"
        >
          Snack or Booze
        </a>
        <ul
          class="ml-auto navbar-nav"
        >
          <li
            class="nav-item"
          >
            <a
              href="/snacks"
            >
              Snacks
            </a>
          </li>
          <li
            class="nav-item"
          >
            <a
              href="/drinks"
            >
              Drinks
            </a>
          </li>
          <li
            class="nav-item"
          >
            <a
              href="/new"
            >
              Add Menu Item
            </a>
          </li>
        </ul>
      </nav>
    </div>
    <main>
      <section
        class="Home"
      >
        <div
          class="card"
        >
          <div
            class="text-center card-body"
          >
            <h5
              class="card-title"
            />
            <h3
              class="font-weight-bold"
            >
              Welcome to Silicon Valley's premier dive cafe!
            </h3>
            <p
              class="card-text"
            />
            <h3>
              We have 
              <a
                class="Home-info"
                href="/snacks"
              >
                3 snacks
              </a>
               and  
              <a
                class="Home-info"
                href="/drinks"
              >
                4 drinks
              </a>
               to choose from!
            </h3>
            <p />
          </div>
        </div>
      </section>
    </main>
  </div>
</DocumentFragment>
`;
